be/src/exec/pipeline/pipeline_driver_queue.h

```diff
@@ -59,16 +73,112 @@ class QuerySharedDriverQueue : public FactoryMethod<DriverQueue, QuerySharedDriv
     static constexpr double RATIO_OF_ADJACENT_QUEUE = 1.2;
     void put_back(const DriverRawPtr driver) override;
     void put_back(const std::vector<DriverRawPtr>& drivers) override;
+
+    void put_back_from_dispatcher(const DriverRawPtr driver) override;
+    void put_back_from_dispatcher(const std::vector<DriverRawPtr>& drivers) override;
+
+    void update_statistics(const DriverRawPtr driver) override;
+
     // return nullptr if queue is closed;
-    StatusOr<DriverRawPtr> take(size_t* queue_index) override;
-    SubQuerySharedDriverQueue* get_sub_queue(size_t) override;
+    StatusOr<DriverRawPtr> take(int dispatcher_id) override;
+
+    size_t size() override;
 
 private:
     SubQuerySharedDriverQueue _queues[QUEUE_SIZE];
     std::mutex _global_mutex;
     std::condition_variable _cv;
-    bool _is_closed;
+    bool _is_closed = false;
+};
+
+// All the QuerySharedDriverQueueWithoutLock's methods MUST be guarded by the outside lock.
+class QuerySharedDriverQueueWithoutLock : public FactoryMethod<DriverQueue, QuerySharedDriverQueueWithoutLock> {
+    friend class FactoryMethod<DriverQueue, QuerySharedDriverQueueWithoutLock>;
+
+public:
+    QuerySharedDriverQueueWithoutLock() {
+        double factor = 1;
+        for (int i = QUEUE_SIZE - 1; i >= 0; --i) {
+            // initialize factor for every sub queue,
+            // Higher priority queues have more execution time,
+            // so they have a larger factor.
+            _queues[i].factor_for_normal = factor;
+            factor *= RATIO_OF_ADJACENT_QUEUE;
+        }
+    }
+    ~QuerySharedDriverQueueWithoutLock() override = default;
+    void close() override {}
+
+    void put_back(const DriverRawPtr driver) override;
+    void put_back(const std::vector<DriverRawPtr>& drivers) override;
+    void put_back_from_dispatcher(const DriverRawPtr driver) override;
+    void put_back_from_dispatcher(const std::vector<DriverRawPtr>& drivers) override;
+
+    // return nullptr if queue is closed;
```

Comment:
What's the return value if queue is closed? Why not a `Status::QueueClosed()` ?

Reply:
This comment already exists before this PR, and is a little misleading. Let me update this comment.
`QuerySharedDriverQueue::take()` and `DriverQueueWithWorkGroup` returns `Status::Cancelled("Shutdown")`, if queue is closed. And `QuerySharedDriverQueueWithoutLock::take()` always return non-nullable value.